Source/Core/InputCommon/ControllerInterface/MappingCommon.cpp

```diff
@@ -163,4 +169,125 @@ bool ContainsCompleteDetection(const Core::InputDetector::Results& results)
   });
 }
 
+ReshapableInputMapper::ReshapableInputMapper(const Core::DeviceContainer& container,
+                                             std::span<const std::string> device_strings)
+{
+  m_input_detector.Start(container, device_strings);
+}
+
+bool ReshapableInputMapper::Update()
+{
+  const auto prev_size = m_input_detector.GetResults().size();
+
+  constexpr auto wait_time = std::chrono::seconds{4};
+  m_input_detector.Update(wait_time, wait_time, wait_time * REQUIRED_INPUT_COUNT);
+
+  return m_input_detector.GetResults().size() != prev_size;
+}
+
+float ReshapableInputMapper::GetCurrentAngle() const
+{
+  constexpr auto quarter_circle = float(MathUtil::TAU) * 0.25f;
+  return quarter_circle - (float(m_input_detector.GetResults().size()) * quarter_circle);
+}
+
+bool ReshapableInputMapper::IsComplete() const
+{
+  return m_input_detector.GetResults().size() >= REQUIRED_INPUT_COUNT ||
+         m_input_detector.IsComplete();
+}
+
+bool ReshapableInputMapper::IsCalibrationNeeded() const
+{
+  return std::ranges::any_of(m_input_detector.GetResults() | std::views::take(REQUIRED_INPUT_COUNT),
+                             &ciface::Core::InputDetector::Detection::IsAnalogPress);
+}
+
+bool ReshapableInputMapper::ApplyResults(ControllerEmu::EmulatedController* controller,
+                                         ControllerEmu::ReshapableInput* stick)
+{
+  auto const detections = m_input_detector.TakeResults();
+
+  if (detections.size() < REQUIRED_INPUT_COUNT)
+    return false;
+
+  // Transpose URDL to UDLR.
+  const std::array results{detections[0], detections[2], detections[3], detections[1]};
+
+  const auto default_device = controller->GetDefaultDevice();
+
+  for (std::size_t i = 0; i != results.size(); ++i)
+  {
+    ciface::Core::DeviceQualifier device_qualifier;
+    device_qualifier.FromDevice(results[i].device.get());
+
+    stick->controls[i]->control_ref->SetExpression(ciface::MappingCommon::GetExpressionForControl(
+        results[i].input->GetName(), device_qualifier, default_device,
+        ciface::MappingCommon::Quote::On));
+
+    controller->UpdateSingleControlReference(g_controller_interface,
+                                             stick->controls[i]->control_ref.get());
+  }
+
```

Comment:
After this loop, please add:

`controller->GetConfig()->GenerateControllerTextures();`